src/particles/PhysicsParticles.hpp

```diff
@@ -0,0 +1,365 @@
+	{
+		if (auto *container = dynamic_cast<quokka::CICParticleContainer *>(this->neighborParticleContainer_)) {
+			amrex::ParticleToMesh(*container, radEnergySource, lev, RadDeposition{current_time, lumIndex, 0, nGroups, birthTimeIndex}, false);
+		}
+	}
+
+	void depositMass(amrex::Vector<amrex::MultiFab> &rhs, int finest_lev, amrex::Real Gconst, int massIndex) override
+	{
+		if (auto *container = dynamic_cast<quokka::CICParticleContainer *>(this->neighborParticleContainer_)) {
+			amrex::ParticleToMesh(*container, amrex::GetVecOfPtrs(rhs), 0, finest_lev, MassDeposition{Gconst, massIndex, 0, 1}, true);
+		}
+	}
+};
+
+// Derived class for CICRad particles
+template <typename problem_t> class CICRadParticleDescriptor : public PhysicsParticleDescriptor
+{
+      public:
+	using PhysicsParticleDescriptor::PhysicsParticleDescriptor;
+
+	void redistribute(int lev) override
+	{
+		if (auto *container = dynamic_cast<quokka::CICRadParticleContainer<problem_t> *>(this->neighborParticleContainer_)) {
+			container->Redistribute(lev);
+		}
+	}
+
+	void redistribute(int lev, int ngrow) override
+	{
+		if (auto *container = dynamic_cast<quokka::CICRadParticleContainer<problem_t> *>(this->neighborParticleContainer_)) {
+			container->Redistribute(lev, container->finestLevel(), ngrow);
+		}
+	}
+
+	void writePlotFile(const std::string &plotfilename, const std::string &name) override
+	{
+		if (auto *container = dynamic_cast<quokka::CICRadParticleContainer<problem_t> *>(this->neighborParticleContainer_)) {
+			container->WritePlotFile(plotfilename, name);
+		}
+	}
+
+	void writeCheckpoint(const std::string &checkpointname, const std::string &name, bool include_header) override
+	{
+		if (auto *container = dynamic_cast<quokka::CICRadParticleContainer<problem_t> *>(this->neighborParticleContainer_)) {
+			container->Checkpoint(checkpointname, name, include_header);
+		}
+	}
+
+	void depositRadiation(amrex::MultiFab &radEnergySource, int lev, amrex::Real current_time, int lumIndex, int birthTimeIndex, int nGroups) override
+	{
+		if (auto *container = dynamic_cast<quokka::CICRadParticleContainer<problem_t> *>(this->neighborParticleContainer_)) {
+			amrex::ParticleToMesh(*container, radEnergySource, lev, RadDeposition{current_time, lumIndex, 0, nGroups, birthTimeIndex}, false);
+		}
+	}
+
+	void depositMass(amrex::Vector<amrex::MultiFab> &rhs, int finest_lev, amrex::Real Gconst, int massIndex) override
+	{
+		if (auto *container = dynamic_cast<quokka::CICRadParticleContainer<problem_t> *>(this->neighborParticleContainer_)) {
+			amrex::ParticleToMesh(*container, amrex::GetVecOfPtrs(rhs), 0, finest_lev, MassDeposition{Gconst, massIndex, 0, 1}, true);
+		}
+	}
+};
+
+// Registry for physics particles
+template <typename problem_t> class PhysicsParticleRegister
+{
+      private:
+	std::vector<std::unique_ptr<PhysicsParticleDescriptor>> particles_;
```

Comment:
We might want to get the PhysicsParticleDescriptor from the PhysicsParticleRegister based on its name. This could be done with `std::map<std::string, std::unique_ptr<PhysicsParticleDescriptor>>` instead.

Reply:
Done.